tests/test_publisher.py

```diff
@@ -0,0 +1,98 @@
+import os
+import queue
+import threading
+from itertools import chain
+
+import pytest
+
+from lektor.publisher import _join_pipes_using_select
+from lektor.publisher import _join_pipes_using_threads
+from lektor.publisher import Command
+
+
+def test_Command_triggers_no_warnings(recwarn):
+    # This excercises the issue where publishing via rsync resulted
+    # in ResourceWarnings about unclosed streams.
+
+    # This is essentially how RsyncPublisher runs rsync.
+    with Command(["echo"], env={}) as client:
+        for _ in client:
+            pass
+    del client
+
+    for warning in recwarn.list:
+        print(warning)
+    assert len(recwarn) == 0
```

Comment:
`recwarn.list` is annotated to be a list. If we do `assert recwarn.list == []` I think we could skip the for loop.

Reply:
The print loop is just to display any warnings on stdout to help with diagnostics should the test fail.  (The `assert len(recwarn) == 0` handles the actual check of the test result.)

I can delete the loop if you'd like.  It just that then, when the test fails you get no indication of what warnings were caught (just how many.)

Reply:
That was my point. If you do `assert recwarn.list == []` instead of the length check pytest will print the extra elements in the array.

Reply:
That only sort of works.  The repr of the WarningMessage objects is not very useful, while the stringification of them is.
```
>>> pytest tests/test_publisher.py -vv
============================= test session starts ==============================
platform linux -- Python 3.8.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /home/dairiki/git/web/lektor-packages/lektor/py38/bin/python
cachedir: .pytest_cache
rootdir: /home/dairiki/git/web/lektor-packages/lektor, configfile: pytest.ini
plugins: cov-2.11.1, mock-3.5.1, click-1.0.2
collected 1 item                                                               

tests/test_publisher.py::test_Command_triggers_no_warnings FAILED        [100%]

=================================== FAILURES ===================================
______________________ test_Command_triggers_no_warnings _______________________

recwarn = WarningsRecorder(record=True)

    def test_Command_triggers_no_warnings(recwarn):
        # This excercises the issue where publishing via rsync resulted
        # in ResourceWarnings about unclosed streams.
    
        # This is essentially how RsyncPublisher runs rsync.
        with Command(["echo"], env={}) as client:
            for _ in client:
                pass
        del client
    
>       assert recwarn.list == []
E       assert [<warnings.WarningMessage object at 0x7f9d3488c130>,\n <warnings.WarningMessage object at 0x7f9d3488c1c0>] == []
E         Left contains 2 more items, first extra item: <warnings.WarningMessage object at 0x7f9d3488c130>
E         Full diff:
E           [
E         -  ,
E         +  <warnings.WarningMessage object at 0x7f9d3488c130>,
E         +  <warnings.WarningMessage object at 0x7f9d3488c1c0>,
E           ]

tests/test_publisher.py:14: AssertionError
```

vs

```
>>> pytest tests/test_publisher.py -vv
============================= test session starts ==============================
platform linux -- Python 3.8.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /home/dairiki/git/web/lektor-packages/lektor/py38/bin/python
cachedir: .pytest_cache
rootdir: /home/dairiki/git/web/lektor-packages/lektor, configfile: pytest.ini
plugins: cov-2.11.1, mock-3.5.1, click-1.0.2
collected 1 item                                                               

tests/test_publisher.py::test_Command_triggers_no_warnings FAILED        [100%]

=================================== FAILURES ===================================
______________________ test_Command_triggers_no_warnings _______________________

recwarn = WarningsRecorder(record=True)

    def test_Command_triggers_no_warnings(recwarn):
        # This excercises the issue where publishing via rsync resulted
        # in ResourceWarnings about unclosed streams.
    
        # This is essentially how RsyncPublisher runs rsync.
        with Command(["echo"], env={}) as client:
            for _ in client:
                pass
        del client
    
        for warning in recwarn.list:
            print(warning)
>       assert len(recwarn) == 0
E       assert 2 == 0
E         +2
E         -0

tests/test_publisher.py:16: AssertionError
----------------------------- Captured stdout call -----------------------------
{message : ResourceWarning('unclosed file <_io.BufferedReader name=11>'), category : 'ResourceWarning', filename : '/home/dairiki/git/web/lektor-packages/lektor/tests/test_publisher.py', lineno : 12, line : None}
{message : ResourceWarning('unclosed file <_io.BufferedReader name=13>'), category : 'ResourceWarning', filename : '/home/dairiki/git/web/lektor-packages/lektor/tests/test_publisher.py', lineno : 12, line : None}
```

Reply:
Ok, leave it then.